extensions/amp-bind/0.1/bind-evaluator.js

```diff
@@ -87,41 +70,106 @@ export class BindEvaluator {
    *   errors: !Object<string, !Error>,
    * }}
    */
-  evaluate(scope) {
+  evaluateBindings(scope) {
     /** @type {!Object<string, ./bind-expression.BindExpressionResultDef>} */
     const cache = {};
     /** @type {!Object<string, !Error>} */
     const errors = {};
 
-    this.parsedBindings_.forEach(binding => {
-      const {tagName, property, expression} = binding;
-      const expr = expression.expressionString;
-
+    this.bindings_.forEach(binding => {
+      const {tagName, property, expressionString} = binding;
       // Skip if we've already evaluated this expression string.
-      if (cache[expr] !== undefined || errors[expr]) {
+      if (cache[expressionString] !== undefined || errors[expressionString]) {
         return;
       }
-
-      let result;
-      try {
-        result = binding.expression.evaluate(scope);
-      } catch (error) {
-        errors[expr] = error;
+      const expression = this.expressionCache_[expressionString];
+      if (!expression) {
+        errors[expressionString] = new Error(
+            `Expression "${expressionString} is not cached."`);
+        return;
+      }
+      const {result, error} = this.evaluate_(expression, scope);
+      if (error) {
+        errors[expressionString] = error;
         return;
       }
-
       const resultString = this.stringValueOf_(property, result);
       if (this.validator_.isResultValid(tagName, property, resultString)) {
-        cache[expr] = result;
+        cache[expressionString] = result;
       } else {
-        errors[expr] = new Error(
+        errors[expressionString] = new Error(
             `"${result}" is not a valid result for [${property}].`);
       }
     });
     return {results: cache, errors};
   }
 
   /**
+   * Evaluates and returns a single expression string.
+   * @param {string} expressionString
+   * @param {!Object} scope
+   * @return {{
+   *   result: ./bind-expression.BindExpressionResultDef,
+   *   error: Error,
+   * }}
+   */
+  evaluateExpression(expressionString, scope) {
```

Comment:
I think we should. A single `setState` expression will likely be executed several times.